Drop the dependency of shirou/gopsutil/v3

nickolaev opened the pull request:
This PR removes the dependency of `github.com/shirou/gopsutil/v3` and introduces `github.com/mackerelio/go-osstat`.
The latter is much simpler, lightweight and has less dependencies.

The main dependency changes are in:
 - `pkg/option/config.go` a trivial change with 1:1 mapping in `mackerelio/go-osstat/memory` package
 - `pkg/maglev/config.go` also trivial 1:1 mapping with the new memory package
 - `pkg/loadinfo/loadinfo.go` this is split in 3 changes: trivial memory and average load replacements and a bit more complex process statistics re-implementation
 
 For the purposes of enumerating the processes in `loadinfo.go` we introduce a naive implementation of parsing `/proc` to find the existing PIDs.
 For all other statistics we use the already vendored `github.com/prometheus/procfs`. The big challenge here was to calculate the PID's CPU usage percentage.
 
 The result of all the above is dropping a huge amount of dependencies:
 - `github.com/shirou/gopsutil/v3`
 - `github.com/go-ole/go-ole`
 - `github.com/lufia/plan9stats`
 - `github.com/power-devops/perfstat`
 - `github.com/tklauser/go-sysconf v0.3.11`
 - `github.com/tklauser/numcpus v0.6.0`
 - `github.com/yusufpapurcu/wmi v1.2.3`

Additinally we introduce 'loadinfo' command in the cilium-dbg binary. It was used to verify that the relevant loadinfo changes work and show sane results.
It can also be useful at debug time to quickly check the resource usage.

Fixes: #27635

```release-note
Drop gopsutil dependecy
```

nickolaev committed:
option/config.go: replace gopsutil with go-osstat

go-osstat is a much simpler implementation with less dependencies.
Here it is almost direct replacement.


Signed-off-by: Nikolay Nikolaev <nikolay.nikolaev@isovalent.com>

nickolaev committed:
maglev/config.go: replace gopsutil with go-osstat

go-osstat is a much simpler implementation with less dependencies.
Here it is almost direct replacement.


Signed-off-by: Nikolay Nikolaev <nikolay.nikolaev@isovalent.com>

nickolaev committed:
loadinfo/loadinfo.go: replace gopsutil with go-osstat for mem and swap

go-osstat is a much simpler implementation with less dependencies.
Here we replace only the memory and swap handling.

We introduce toPercent function to calculate usage percent since
go-osstat does not provide these values.


Signed-off-by: Nikolay Nikolaev <nikolay.nikolaev@isovalent.com>

nickolaev committed:
loadinfo/loadinfo.go: replace gopsutil with go-osstat for average load

go-osstat is a much simpler implementation with less dependencies.
Here we replace only the average load handling.

We introduce toPercent function to calculate usage percent since
go-osstat does not provide these values.


Signed-off-by: Nikolay Nikolaev <nikolay.nikolaev@isovalent.com>

nickolaev committed:
cilium-dbg: add loadinfo command

This will just invoke the loadinfo module and show the current
system load as if it would be logged if the datapath debug is on.

Signed-off-by: Nikolay Nikolaev <nikolay.nikolaev@isovalent.com>

derailed commented on pkg/loadinfo/loadinfo.go:
```diff
@@ -36,46 +38,89 @@ func toMB(total uint64) uint64 {
+		return nil, err
+	}
+
+	pids := make([]int, len(matches))
```
len=0? This will append on top of the pre-allocated slice.
Perhaps add a test here?

nickolaev replied:
IIUC `/proc/pid` will always have at least 1 pid accessible to the current user (see hidepid here https://man7.org/linux/man-pages/man5/proc.5.html) so the chances for len to be 0 are 0 :)
But even if `matches` is empty, we will never iterate and try to add anything in pids.
A test is probably a good idea, but it is kind alwasy going to return random number of pids.

tklauser replied:
I think what @derailed meant was that this should `make` a slice with `len=0` and `cap=len(matches)`. Currently, `pids` has `2*len(matches)` entries after the loop below.

nickolaev replied:
Right!

derailed commented on pkg/loadinfo/loadinfo.go:
```diff
@@ -36,46 +38,89 @@ func toMB(total uint64) uint64 {
 
-	procs, err := process.Processes()
+	pids, err := pids()
 	if err == nil {
```
We may want to inverse the logic here and eliminate the indent?

nickolaev replied:
I wanted to stick with the original code structure as much as possible to make it a replacement, not re-write.

derailed commented on pkg/loadinfo/loadinfo.go:
```diff
@@ -36,46 +38,89 @@ func toMB(total uint64) uint64 {
+			}
+
+			cpuPercent := cpuPercent(stat)
 			if cpuPercent > cpuWatermark {
```
Same here short cut the loop if below threshold

nickolaev replied:
Again, I did try to stick with the original code structure. In any case if there is much desire I can change it.

derailed reviewed (changes requested):
@nickolaev Nice work and clean up!

ysksuzuki commented on pkg/maglev/maglev.go:
```diff
@@ -203,7 +204,7 @@ func GetLookupTable(backendsMap map[string]*loadbalancer.Backend, m uint64) []in
 func derivePermutationSliceLen(m uint64) uint64 {
 	threshold := uint64(8 * 1024 * 1024 * 1024) // 8GB
-	if vm, err := mem.VirtualMemory(); err != nil || vm == nil || vm.Total <= threshold {
+	if vm, err := memory.Get(); err != nil || vm == nil || vm.Total <= threshold {
```
Both libraries retrieve MemTotal from `proc/meminfo`. So the result will be the same.

gopsutil
https://github.com/shirou/gopsutil/blob/v3.23.12/mem/mem_linux.go#L59

go-osstat
https://github.com/mackerelio/go-osstat/blob/v0.2.4/memory/memory_linux.go#L34

ysksuzuki reviewed (approved):
sig-lb changes LGTM

derailed reviewed (approved):
@nickolaev LGTM

nickolaev committed:
documentation: update cilium-dbg for loadinfo

Signed-off-by: Nikolay Nikolaev <nikolay.nikolaev@isovalent.com>

learnitall reviewed (approved):
Looks good for docs!

rolinh reviewed (approved):
go-osstat is licensed apache 2.0 and seems to see enough adoption already thus vendor changes lgtm

nickolaev committed:
loadinfo/loadinfo.go: replace gopsutil with procfs for processes dump

prometheus/procfs is already a dependency and provides the basic
/proc parsing functionality that we need. This should be enough
for the purposes of this file.

We introduce two new functions:
 - pids() to fetch a slice of the current pids available on the system
 - cpuPercent to calculate the process CPU usage percentage

Running `go mod tidy && go mod vendor` cleans up a huge amount of now
unneed dependencies.

Signed-off-by: Nikolay Nikolaev <nikolay.nikolaev@isovalent.com>
